Adding optional revision bump to snapshot restore

dusk125 opened the pull request:
This PR adds the flags `--bump-revision` and `--mark-compacted` to etcdutl snapshot restore for #16028.

I chose to handle this after the database is loaded from the snapshot, so as to not modify the snapshot, I didn't want to risk corrupting the snapshot as a part of this optional step. If for some reason the bump fails, a user would still be able to restore the snapshot again without the bump and restart the watchers manually.

serathius commented:
To clarify by `bumping revision` I didn't mean take transform whole change history by shifting revisions of all modifications, but to just bump revision of the next transaction.

serathius commented on etcdutl/etcdutl/snapshot_command.go:
```diff
@@ -75,6 +76,7 @@ func NewSnapshotRestoreCommand() *cobra.Command {
 	cmd.Flags().StringVar(&restorePeerURLs, "initial-advertise-peer-urls", defaultInitialAdvertisePeerURLs, "List of this member's peer URLs to advertise to the rest of the cluster")
 	cmd.Flags().StringVar(&restoreName, "name", defaultName, "Human-readable name for this member")
 	cmd.Flags().BoolVar(&skipHashCheck, "skip-hash-check", false, "Ignore snapshot integrity hash value (required if copied from data directory)")
+	cmd.Flags().Int64Var(&revisionBump, "bump-revision", 0, "How much to increase the revision of each key, after restore, to force a resync of watchers")
```
```suggestion
	cmd.Flags().Int64Var(&revisionBump, "bump-revision", 0, "How much to increase the revision of the next etcd transaction ")
```

dusk125 commented:
With this latest update, I've changed the revision-bump flag to only increase the latest revision by the amount so that clients shouldn't see decreasing revisions.
Also, I added the other flag, mark-compacted, to set the latest revision as a scheduled "in progress" compaction so that watchers will probably fail and resync.

serathius commented on etcdutl/snapshot/util.go:
```diff
@@ -20,12 +20,23 @@ import (
 	sub  int64
 }
 
+// The binary representation of the revision is 17 bytes:
```
Dob't think that's true. I remember that there might be 18byte to mark a tombstone. I will need to confirm, but we should not copy the code to parsing/dumping revisions lightly.

serathius replied:
https://github.com/etcd-io/etcd/blob/7444985cab75e3d8ef4ef5ffe71634547556aa68/server/storage/mvcc/kvstore.go#L523-L532

serathius commented on etcdutl/snapshot/v3_snapshot.go:
```diff
@@ -311,6 +333,79 @@ func (s *v3Manager) saveDB() error {
+// Compaction should be run after this to further help invalidate watcher caches.
+func (s *v3Manager) revBump(amount int64) error {
+	// Don't do anything if there's nothing to do
+	if amount == 0 {
```
Such checks should be outside of function or function should be called `maybeRevBump` as it might not do the bump.

serathius commented on etcdutl/snapshot/v3_snapshot.go:
```diff
@@ -311,6 +333,79 @@ func (s *v3Manager) saveDB() error {
+// 'amount' should be a very large number, > 1_000_000.
+//
+// Compaction should be run after this to further help invalidate watcher caches.
+func (s *v3Manager) revBump(amount int64) error {
```
```suggestion
func (s *v3Manager) bumpRevision(amount int64) error {
```

serathius commented on etcdutl/snapshot/util.go:
```diff
@@ -20,12 +20,23 @@ import (
+// The binary representation of the revision is 17 bytes:
+//
+// [8 bytes: main] [1 byte: sep] [8 bytes: sub]
 func bytesToRev(bytes []byte) revision {
```
Don't like the idea of copying this code. It's too fragile. Still making this public might not be best solution. Can you  comment that `bytesToRev` and `revToBytes` functions should be synced with function server.

serathius commented on etcdutl/snapshot/v3_snapshot.go:
```diff
@@ -311,6 +333,79 @@ func (s *v3Manager) saveDB() error {
+	defer be.Close()
+
+	tx := be.BatchTx()
+	tx.LockInsideApply()
```
This is not apply function so `LockOutsideApply`

serathius commented on etcdutl/snapshot/v3_snapshot.go:
```diff
@@ -311,6 +333,79 @@ func (s *v3Manager) saveDB() error {
+	tx.LockInsideApply()
+	defer tx.Unlock()
+
+	var latest revision
+	var value []byte
+	if err := tx.UnsafeForEach(schema.Key, func(k, v []byte) (err error) {
+		rev := bytesToRev(k)
+
+		if latest.main < rev.main {
+			latest = rev
+			value = v
+		}
+
+		return nil
+	}); err != nil {
+		return err
+	}
```
Move to separate function.

serathius commented on etcdutl/snapshot/v3_snapshot.go:
```diff
@@ -311,6 +333,79 @@ func (s *v3Manager) saveDB() error {
+	tx.LockInsideApply()
+	defer tx.Unlock()
+
+	var latest int64
+	if err := tx.UnsafeForEach(schema.Key, func(k, v []byte) (err error) {
+		rev := bytesToRev(k)
+
+		if latest < rev.main {
+			latest = rev.main
+		}
+
+		return nil
+	}); err != nil {
+		return err
+	}
```
This code repeats, please move to separate function.

serathius commented on etcdutl/snapshot/v3_snapshot.go:
```diff
@@ -311,6 +333,79 @@ func (s *v3Manager) saveDB() error {
+// revBump increases the revision, modrevision, and version of every key to force watchers to resync.
+//
+// In order to ensure that watchers don't have a revision that's higher than the restored revision,
+// 'amount' should be a very large number, > 1_000_000.
```
1`000`000 is not a large number for revision. Restored backup could have couple of days. 1 QPS for 11 days is enough to hit 1000000. Please suggest at least 1`000`000`000, even with such large bump it's enough to make 9`223`372`036 backups.

dusk125 commented on etcdutl/snapshot/v3_snapshot.go:
```diff
@@ -311,6 +333,79 @@ func (s *v3Manager) saveDB() error {
+	defer be.Close()
+
+	tx := be.BatchTx()
+	tx.LockInsideApply()
```
Looking at [mvcc.SetScheduledCompact](https://github.com/etcd-io/etcd/blob/main/server/storage/mvcc/store.go#L38), it does a LockInsideApply then calls `UnsafeSetScheduledCompact`. So for getting the latest revision I don't need LockInsideApply, but for setting the schedule, it looks like I should. 
Should I have two different transactions or just stick with InsideApply?

serathius commented on etcdutl/snapshot/util.go:
```diff
@@ -20,12 +20,27 @@ import (
 	sub  int64
 }
 
+// The binary representation of the revision is 17 bytes
+// (sometimes 18 bytes see mvcc.isTombstone)
+//
+// [8 bytes: main] [1 byte: sep] [8 bytes: sub]
```
Please don't document this, especially here. Don't want to spent too much time on getting this right when we just want to fix restore function. If you want to improve function documentation please create a separate PR.

serathius commented on etcdutl/snapshot/util.go:
```diff
@@ -20,12 +20,27 @@ import (
+//
+// [8 bytes: main] [1 byte: sep] [8 bytes: sub]
+//
+// TODO bytesToRev should be synced with function server
```
Please provide a proper link to the bytesToRev function in server

serathius commented on etcdutl/snapshot/util.go:
```diff
@@ -20,12 +20,27 @@ import (
 	}
 }
+
+// TODO revToBytes should be synced with function server
```
ditto

serathius commented on etcdutl/snapshot/v3_snapshot.go:
```diff
@@ -203,6 +204,16 @@ type RestoreConfig struct {
 	SkipHashCheck bool
+
+	// RevisionBump is the amount to increase the latest revision after restore,
+	// to ensure that the revisions never decrease from the clients perspective
```
```suggestion
	// to allow administrators to trick clients into thinking that revision never decreased.
```

serathius replied:
bump revision doesn't ensure anything, it's up to administrators to provide RevisionBump large enough to trick users.

serathius commented on etcdutl/snapshot/v3_snapshot.go:
```diff
@@ -203,6 +204,16 @@ type RestoreConfig struct {
+	// If 0, revision bumping is skipped.
+	RevisionBump int64
+
+	// MarkCompacted is "true" to mark the latest revision as scheduled for compaction.
```
```suggestion
	// MarkCompacted is "true" to mark the latest revision as compacted.
```

serathius commented on etcdutl/snapshot/v3_snapshot.go:
```diff
@@ -203,6 +204,16 @@ type RestoreConfig struct {
+	RevisionBump int64
+
+	// MarkCompacted is "true" to mark the latest revision as scheduled for compaction.
+	// This will cause the server to compact the restored revisions (at some point after startup),
```
```suggestion
	// This will cause the server to compact the restored revisions (at some point after startup),
```
This is implementation detail.

serathius commented on etcdutl/snapshot/v3_snapshot.go:
```diff
@@ -203,6 +204,16 @@ type RestoreConfig struct {
+
+	// MarkCompacted is "true" to mark the latest revision as scheduled for compaction.
+	// This will cause the server to compact the restored revisions (at some point after startup),
+	// and correctly cancel watchers so they can invalidate their caches.
```
Comments are nice, but it would be better to just provide all this information as part of user visible information in flag help section.

tjungblu replied:
as commented earlier, I think this is too focused on K8s as a consumer. In both places, I would leave that information out.

serathius commented on etcdutl/snapshot/v3_snapshot.go:
```diff
@@ -311,6 +335,88 @@ func (s *v3Manager) saveDB() error {
 }
 
+// unsafeGetLatestRevision returns the latest revision without locking the given transaction.
+func (s *v3Manager) unsafeGetLatestRevision(tx backend.BatchTx) (latest revision, err error) {
```
Code is read from top to bottom, please move the function to after `markLatestRevCompacted` as it's the least important one.

serathius commented on etcdutl/snapshot/v3_snapshot.go:
```diff
@@ -311,6 +335,88 @@ func (s *v3Manager) saveDB() error {
 	return nil
 }
 
+// unsafeGetLatestRevision returns the latest revision without locking the given transaction.
```
```suggestion
```

serathius replied:
Whole comment is nicely reflected in the function name. Comments get outdated.

serathius commented on etcdutl/snapshot/v3_snapshot.go:
```diff
@@ -311,6 +335,88 @@ func (s *v3Manager) saveDB() error {
+
+// bumpRevision increases the revision so that, from the client's perspective, the revision never decreases.
+//
+//	If we don't bump the revision, a client looking for a key at one revision finds that the key is actually at a
```
Don't put feature documentation within a small function. This documentation should be user facing not be a comment that no user will see.

serathius commented on etcdutl/snapshot/v3_snapshot.go:
```diff
@@ -311,6 +335,88 @@ func (s *v3Manager) saveDB() error {
+//	smaller revision which breaks the promise that revisions never decrease. The client has no way of knowing that
+//	etcd was restored, it only sees the decrease in revision.
+//
+// To ensure that clients don't have a revision that's higher than the restored revision,
```
ditto.

serathius commented on etcdutl/snapshot/v3_snapshot.go:
```diff
@@ -311,6 +335,88 @@ func (s *v3Manager) saveDB() error {
+// To ensure that clients don't have a revision that's higher than the restored revision,
+// 'amount' should be a very large number, >= 1_000_000_000.
+//
+// Compaction should be run after this to further help invalidate watcher caches.
```
This is incorrect behavior as those flags are independent. Please implement bumpRevision in a way that doesn't require `markLatestRevCompacted` or add a validation that forbids using `bumpRevision` without using `markLatestRevCompacted`.

chaochn47 replied:
+1. Sounds like `bumpRevision` + `markLatestRevCompacted` should be used together to mark user last observed revision is compacted.

dusk125 replied:
Agreed that if we bumpRevision, we want to compact. Can either of you foresee wanting to compact without bumping: should that be an option?

serathius commented on etcdutl/snapshot/v3_snapshot.go:
```diff
@@ -311,6 +335,88 @@ func (s *v3Manager) saveDB() error {
+// Compaction should be run after this to further help invalidate watcher caches.
+// (see markLatestRevCompacted)
+func (s *v3Manager) bumpRevision(amount int64) error {
+	be := backend.NewDefaultBackend(s.lg, s.outDbPath())
```
Please open the backed only once.

serathius commented on etcdutl/snapshot/v3_snapshot.go:
```diff
@@ -311,6 +335,88 @@ func (s *v3Manager) saveDB() error {
+// (see markLatestRevCompacted)
+func (s *v3Manager) bumpRevision(amount int64) error {
+	be := backend.NewDefaultBackend(s.lg, s.outDbPath())
+	defer be.Close()
```
Please call `ForceCommit` before close.

serathius commented on etcdutl/snapshot/v3_snapshot.go:
```diff
@@ -311,6 +335,88 @@ func (s *v3Manager) saveDB() error {
+// markLatestRevCompacted sets the scheduled compaction to the latest revision in order to tell the server
+// that it should compact the restored database up to the latest revision.
+//
+// This will cause any watcher to be cancelled as the revision they're watching will be compacted,
```
This is restore feature documentation, not a good much for a low level function.

serathius commented on etcdutl/snapshot/v3_snapshot.go:
```diff
@@ -265,6 +276,19 @@ func (s *v3Manager) Restore(cfg RestoreConfig) error {
 		return err
 	}
+
+	if cfg.RevisionBump > 0 {
```
Both bumpRevision and markLatestRevCompacted should be done in one transaction so it can be reverted on error.

serathius commented:
Please add e2e test.

tjungblu commented on etcdutl/snapshot/v3_snapshot.go:
```diff
@@ -265,6 +276,19 @@ func (s *v3Manager) Restore(cfg RestoreConfig) error {
 		return err
 	}
+
+	if cfg.RevisionBump > 0 {
```
I think `!= 0` would be more idiomatic. I'm not sure if we even want to allow to pass negative numbers in the first place?

dusk125 replied:
Yeah, I can't think of a reason that we would want to allow someone to decrease the revision, especially since the bump is there to help make sure that doesn't happen.

tjungblu commented on etcdutl/snapshot/v3_snapshot.go:
```diff
@@ -311,6 +335,88 @@ func (s *v3Manager) saveDB() error {
+	)
+
+	latest.main += amount
+	k := [17]byte{}
```
also a good place to rather make the revision.go public and reuse `revBytesLen`

tjungblu commented on etcdutl/snapshot/v3_snapshot.go:
```diff
@@ -311,6 +335,88 @@ func (s *v3Manager) saveDB() error {
+	}
+
+	s.lg.Info(
+		"bumping latest revision to ensure revisions never decrease",
```
I think a short statement of `bumping latest revision` is already enough, you don't need to justify everything  :)

serathius commented on etcdutl/snapshot/util.go:
```diff
@@ -20,12 +20,27 @@ import (
 }
+
+// TODO revToBytes should be synced with function server
+func revToBytes(bytes []byte, rev revision) {
```
It's out of scope.

dusk125 replied:
I can handle that in another PR so the refactor doesn't muddy this PR, it would be nice to remove the duplication.

serathius commented on etcdutl/snapshot/v3_snapshot.go:
```diff
@@ -311,6 +327,73 @@ func (s *v3Manager) saveDB() error {
 }
 
+// maybeModifyLatestRevision can increase the latest revision by the given amount and sets the scheduled compaction
+// to that revision so that the server will compact the database up to that revision.
```
```suggestion
// to that revision so that the server will consider this revision compacted.
```

serathius commented on etcdutl/snapshot/util.go:
```diff
@@ -20,12 +20,24 @@ import (
 
 type revision struct {
 	main int64
+	sep  byte
```
Why add `sep`?  Not needed

serathius replied:
This struct should be kept synched with revision from https://github.com/etcd-io/etcd/blob/main/server/storage/mvcc/revision.go

serathius commented on etcdutl/snapshot/v3_snapshot.go:
```diff
@@ -311,6 +327,73 @@ func (s *v3Manager) saveDB() error {
+	err := tx.UnsafeForEach(schema.Key, func(k, _ []byte) (err error) {
+		rev := bytesToRev(k)
+
+		if latest.main < rev.main {
```
This is not the latest revision as you are not considering subrevision. Either expose main revision as int64 or implement a proper `unsafeGetLatestRevision`

serathius commented on etcdutl/snapshot/v3_snapshot.go:
```diff
@@ -311,6 +327,73 @@ func (s *v3Manager) saveDB() error {
+		zap.Int64("revision", latest.main),
+	)
+
+	mvcc.UnsafeSetScheduledCompact(tx, latest.main)
```
Please make sure to get semantics of subrevision right. I think to mark revision as compacted (X,Y) we put (X+1, 0) as compacted.

dusk125 replied:
Do you mean that you would put an additional revision at X+1 then pass that into SetScheduledCompact or pass in X+1 to SetScheduledCompact? The latter at least doesn't work and causes a future revision error on startup.

ptabor replied:
Wasn't this about watch ?

serathius commented on etcdutl/snapshot/v3_snapshot.go:
```diff
@@ -311,6 +327,73 @@ func (s *v3Manager) saveDB() error {
+		zap.Int64("new-latest-revision", latest.main+amount),
+	)
+
+	latest.main += amount
```
Please make sure to get semantics of subrevision right. I think to bumping revision (X,Y) by N should result in revision (X+N, 0).

dusk125 commented:
That last test failure was a flake, sometimes the watch would take a tad too long to start and miss the first put. I've fixed this in the latest iteration with a short sleep after starting the watch: I ran the current iteration a couple thousand times with no more flakes (beforehand it was at least 1 in 100 flaky).

serathius commented on tests/e2e/ctl_v3_snapshot_test.go:
```diff
@@ -305,3 +308,126 @@ func snapshotVersionTest(cx ctlCtx) {
+	cmdPrefix := []string{e2e.BinPath.Etcdctl, "--endpoints", strings.Join(epc.EndpointsGRPC(), ",")}
+	require.NoError(t, e2e.SpawnWithExpects(append(cmdPrefix, "snapshot", "save", fpath), nil, fmt.Sprintf("Snapshot saved at %s", fpath)))
+
+	t.Log("Stopping the original server...")
```
Please add some writes after snapshot but before the stopping of the server. We should test that writes after snapshot are lost and watch correctly behaved thanks to revision bump.

serathius commented on tests/e2e/ctl_v3_snapshot_test.go:
```diff
@@ -305,3 +308,126 @@ func snapshotVersionTest(cx ctlCtx) {
+	}, "added member")
+	require.NoError(t, err)
+
+	t.Log("(Re)starting the etcd member using the restored snapshot...")
+	epc.Procs[0].Config().DataDirPath = newDataDir
+	for i := range epc.Procs[0].Config().Args {
+		if epc.Procs[0].Config().Args[i] == "--data-dir" {
+			epc.Procs[0].Config().Args[i+1] = newDataDir
+		}
+	}
```
Why not delete old data directory and restore into it?

tjungblu replied:
to be fair, I've copy&pasted from the testcase above this one yesterday. I assume it was easier to debug the older data if anything goes wrong, but can't tell what the original authors intention here was.

serathius replied:
Ok, maybe let's have this as a followup to avoid inconsistent code.

serathius commented on tests/e2e/ctl_v3_snapshot_test.go:
```diff
@@ -305,3 +308,126 @@ func snapshotVersionTest(cx ctlCtx) {
+		require.Equal(t, int64(1), v.Kvs[0].Version)
+	}
+
+	cancelResult := <-watchCh
```
Can you also test that `watchCh` is closed?

serathius commented on tests/e2e/ctl_v3_snapshot_test.go:
```diff
@@ -305,3 +308,126 @@ func snapshotVersionTest(cx ctlCtx) {
+		require.Equal(t, int64(1), v.Count)
+		require.Equal(t, kvs[i].Val, string(v.Kvs[0].Value))
+		// the currentRev we had before the snapshot should be bumped by now
+		require.Equal(t, int64(bumpAmount+currentRev), v.Header.Revision)
```
Testing header revision is not important here. Aside of that data should match one from before, so we could refactor it into separate function.

tjungblu replied:
it's not important, but I don't really see how those assertions are hurting. Good point on the function though, @dusk125 do you want to take this on from here? I need to start working on the other stuff today.

serathius commented on tests/e2e/ctl_v3_snapshot_test.go:
```diff
@@ -305,3 +308,126 @@ func snapshotVersionTest(cx ctlCtx) {
+
+	require.NoError(t, epc.Restart(context.Background()))
+
+	t.Log("Ensuring the restored member has the correct data...")
```
Please add validation that user can no longer request old revisions from before restore. Like revisions that were lost.

tjungblu reviewed (approved):
/lgtm (nonbinding)

serathius commented:
cc @ahrtr @ptabor @spzala @mitake

ptabor commented on tests/framework/e2e/etcdctl.go:
```diff
@@ -698,6 +698,9 @@ func (ctl *EtcdctlV3) Watch(ctx context.Context, key string, opts config.WatchOp
 					json.Unmarshal([]byte(line), &resp)
 					if resp.Canceled {
+						if opts.IncludeCancellationResponse {
+							ch <- resp
```
Why this change. Please add a comment .

tjungblu replied:
the cancellation response was swallowed before and existing tests are not expecting this result. We need to assert that it is returned in: 
https://github.com/etcd-io/etcd/pull/16029/files#diff-d99fed41bb3f2962001bb422375c286f00cf1e1e9f57d8bd596f11bde187dea7R399-R405

serathius replied:
I think this is because etcdctl doesn't print cancel response, just exists.

serathius replied:
Oh, forgot to comment on this. Not passing cancel response is just oversight when implementing etcdctl struct. This should happen by default. 

@tjungblu Have you tested just passing it by default? Does it cause other test to fail?

dusk125 replied:
Doesn't look like making that the default broke anything, I'll make this change

dusk125 commented on etcdutl/snapshot/v3_snapshot.go:
```diff
@@ -311,6 +327,75 @@ func (s *v3Manager) saveDB() error {
+	latest.main += amount
+	latest.sub = 0
+	k := [17]byte{}
+	revToBytes(k[:], latest)
+	tx.UnsafePut(schema.Key, k[:], []byte{})
```
Because k is a fixed-sized array and both functions take in a slice

ahrtr replied:
Got it, then I would suggest to change line 369 to `k := make([]byte, 17)`.  Your current code works, but it's less readable.

dusk125 commented on etcdutl/snapshot/v3_snapshot.go:
```diff
@@ -265,6 +274,13 @@ func (s *v3Manager) Restore(cfg RestoreConfig) error {
 		return err
 	}
+
+	if cfg.MarkCompacted || cfg.RevisionBump > 0 {
```
I'm not sure that's necessarily the case; we need to compact if we're bumping the revision, but if we're not bumping the revision, the user may still want to mark everything in the restore as compacted. That was at least my understanding as to why to have them be separate options. @serathius thoughts?

ahrtr replied:
> but if we're not bumping the revision, the user may still want to mark everything in the restore as compacted

Why? What this PR does is already a little hacky, why do we need to provide more flexibility? I do not see any real use case, but it provides flexibility for users to make mistake.

serathius replied:
Separate flags for clarity, but I agree with @ahrtr that we don't need to make them independent. It would only require more work to test all edge cases.

ahrtr commented on etcdutl/etcdutl/snapshot_command.go:
```diff
@@ -38,6 +38,8 @@ var (
 	restoreName         string
 	skipHashCheck       bool
+	markCompacted       bool
+	revisionBump        int64
```
You do not make the suggested change, nor add any comment, why do you resolve the comment directly?

dusk125 replied:
I have been marking simple changes like this resolved as I handle them locally, so I can keep track of what I've fixed. This, and the below, changes will be in my next update.

ahrtr commented on etcdutl/etcdutl/snapshot_command.go:
```diff
@@ -75,6 +77,8 @@ func NewSnapshotRestoreCommand() *cobra.Command {
 	cmd.Flags().StringVar(&restorePeerURLs, "initial-advertise-peer-urls", defaultInitialAdvertisePeerURLs, "List of this member's peer URLs to advertise to the rest of the cluster")
 	cmd.Flags().StringVar(&restoreName, "name", defaultName, "Human-readable name for this member")
 	cmd.Flags().BoolVar(&skipHashCheck, "skip-hash-check", false, "Ignore snapshot integrity hash value (required if copied from data directory)")
+	cmd.Flags().Int64Var(&revisionBump, "bump-revision", 0, "How much to increase the latest revision after restore")
```
ditto

ahrtr commented on etcdutl/etcdutl/snapshot_command.go:
```diff
@@ -110,12 +114,19 @@ func SnapshotRestoreCommandFunc(restoreCluster string,
 		cobrautl.ExitWithError(cobrautl.ExitBadArgs, err)
 	}
 
+	if (revisionBump == 0 && markCompacted) || (revisionBump > 0 && !markCompacted) {
```
Do we support only bumping revision, and not marking compacted?

dusk125 replied:
No, we only support bumping revision and marking compacted or not bumping revision and not marking compacted. As Marek said above, the flags are separate for clarity, but they need to be used together.

dusk125 replied:
@serathius thoughts?

serathius replied:
Don't understand the argument presented by @ahrtr, I think we are getting into non important details of personal preference. I would recommend to keep the flags separate as it's more clear to user and it's easier later to remove a flag then split the flag if we would decide to support bumping revision without marking it as compacted.

ahrtr replied:
Don't think any user-facing change is non-important. Marking compact should never be a must. I am evening think why do we need it on second thought. Please make sure you understand my comment above and let me know if you have confusion.

serathius replied:
Please read and comment on the original proposal https://github.com/kubernetes/kubernetes/issues/118501, not spread the discussion and design options in just before merging the code just because of "second thought". If we need to change the design, then let's discuss changing the design.

ahrtr replied:
There is no change on the design per my previous comment https://github.com/etcd-io/etcd/pull/16029#discussion_r1243910953.  The reason for raising the comment is to make sure we have thought about other possibilities.

This PR fall into the case 2.2 mentioned above, which I don't strongly disagree. If we decide to follow this way, then we may consider removing "`--mark-compact`", and do it (mark compaction) automatically when bumping the revision. Obviously it's clearer & easier to users.

serathius replied:
The design https://github.com/kubernetes/kubernetes/issues/118501 proposed your case 2.2, two points `Etcd restore operation should support bumping revision by an arbitrary number` and `
Etcd restore operation should support marking old revisions as compacted` are separated and are not a single point. Meaning those are independent.

ahrtr replied:
Unfortunately we don't have enough active maintainers and/or reviewers. Each user facing change or etcd core change should have enough reviews and feedback beforing rejecting or merging.

For this case, let's go for 2.2, and let's be aware that  `The only purpose is to make the users be explicitly aware of the operation` . Otherwise, when we are challenged by users, at least we know how to respond.

serathius replied:
Ok, could we clarify what the action for @dusk125 ? Is this comment addressed, or they should change something?

dusk125 commented:
I think that test failure may be a flake, I'm retesting locally to see if it'll reproduce

ahrtr reviewed (approved):
LGTM

Thanks @dusk125

ahrtr commented:
Two left actions:
- Please raise a followup ticket for moving some revision related common functions to package "`pkg`", see https://github.com/etcd-io/etcd/pull/16029#discussion_r1242315620
- Please raise a followup ticket for documentation for the flag "--mark-compact". 
  - It isn't allowed to mark compact when users don't bump revision;
  - If users bump revision, then users must specify the flag. The purpose is to make sure users are explicitly aware of the operation (mark compaction). (Note that if we receive too many negative feedback from users, we may consider to remove the flag, and go for 2.1 in future).

dusk125 commented:
> Two left actions:
> 
> * Please raise a followup ticket for moving some revision related common functions to package "`pkg`", see [Adding optional revision bump to snapshot restore #16029 (comment)](https://github.com/etcd-io/etcd/pull/16029#discussion_r1242315620)
> * Please raise a followup ticket for documentation for the flag "--mark-compact".
>   
>   * It isn't allowed to mark compact when users don't bump revision;
>   * If users bump revision, then users must specify the flag. The purpose is to make sure users are explicitly aware of the operation (mark compaction). (Note that if we receive too many negative feedback from users, we may consider to remove the flag, and go for 2.1 in future).

Tickets have been raised

tjungblu commented:
While @dusk125 is out, I can take over the 3.5 backport. Can anyone hit the merge button? Thanks.
